operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1797 @@
+	ipv4Enabled bool
+	ipv6Enabled bool
+
+	poolClient cilium_client_v2alpha1.CiliumLoadBalancerIPPoolInterface
+	svcClient  client_typed_v1.ServicesGetter
+
+	poolResource resource.Resource[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+	svcResource  resource.Resource[*slim_core_v1.Service]
+
+	shutdowner hive.Shutdowner
+
+	poolStore    poolStore
+	rangesStore  rangesStore
+	serviceStore serviceStore
+
+	ctx    context.Context
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	// Only used during testing.
+	initDoneCallbacks []func()
+}
+
+// Start implements hive.HookInterface
+func (ipam *LBIPAM) Start(_ hive.HookContext) error {
+	ipam.logger.Info("Starting LB IPAM")
+
+	ipam.wg.Add(1)
+
+	// Start go routine, we can't block in OnStart
+	go func() {
+		defer ipam.wg.Done()
+		ipam.Run()
+	}()
+
+	return nil
+}
+
+// Stop implements hive.HookInterface
+func (ipam *LBIPAM) Stop(ctx hive.HookContext) error {
+	// Close the ctx
+	ipam.cancel()
+
+	// Convert blocking Wait into a channel so we can select on the context closing or the Wait.
+	wgDone := make(chan struct{})
+	go func() {
+		ipam.wg.Wait()
+		close(wgDone)
+	}()
+
+	select {
+	case <-wgDone:
+	case <-ctx.Done():
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) restart() {
+	ipam.logger.Info("Restarting LB IPAM")
+
+	// Refresh the context
+	ipam.ctx, ipam.cancel = context.WithCancel(context.Background())
+
+	// Reset all stored state
+	ipam.poolStore = NewPoolStore()
+	ipam.rangesStore = newRangesStore()
+	ipam.serviceStore = NewServiceStore()
+
+	// Re-start the main goroutine
+	ipam.wg.Add(1)
+	go func() {
+		defer ipam.wg.Done()
+		ipam.Run()
+	}()
+}
+
+type ipPoolEvent = resource.Event[*cilium_api_v2alpha1.CiliumLoadBalancerIPPool]
+type svcEvent = resource.Event[*slim_core_v1.Service]
+
+func (ipam *LBIPAM) Run() {
+	// PreInit blocks and waits until there is at least 1 pool, which is the trigger to activate the controller.
+	ipam.logger.Info("LB-IPAM in pre-init phase, waiting for pools")
+	err := ipam.preInit()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error while pre-initializing LB-IPAM")
+		return
+	}
+	// If we exited preInit because we are shutting down, just return
+	if ipam.ctx.Err() != nil {
+		return
+	}
+
+	errChan := make(chan error, 2)
+	poolChan := stream.ToChannel[ipPoolEvent](ipam.ctx, errChan, ipam.poolResource)
+	svcChan := stream.ToChannel[svcEvent](ipam.ctx, errChan, ipam.svcResource)
+
+	ipam.logger.Info("LB-IPAM initializing")
+	err = ipam.init(ipam.ctx, poolChan, svcChan)
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error while initializing LB-IPAM")
+		// Stop, don't go into main control loop
+		return
+	}
+
+	ipam.logger.Info("LB-IPAM done initializing")
+	for _, cb := range ipam.initDoneCallbacks {
+		if cb != nil {
+			cb()
+		}
+	}
+
+	// When the ctx expires, both streams will close, keep processing until both channels are fully drained to avoid
+	// blocking the resources in case they are re-used after we shutdown.
```

Comment:
IMO we shouldn't allow re-usage after shutting down.

Reply:
@joamaki suggested that I make it like this to ensure I leave the Resource[T] in a good state, I believe it has to do with no leaking goroutines that are blocking while testing ect. Perhaps the comment is a bit misleading in that case.

Reply:
Yeah the point here is that we always need to consume the streams to the end, even if we cancel. If we don't, we may end up blocking them [1], which at the very least will prevent graceful termination if there are other observers for the same resource. Reword the comment to not talk about re-use, but just mention that this is done to drain in order to not block.

[1]: In case of `Resource[T]` we have a goroutine pulling from a work queue and calling the `next` function on each of the items. If we use `ToChannel` that `next` function will be sending to a channel. If we don't drain, then that function blocks on trying to send to the (by default unbuffered) channel. Thus any use of `ToChannel` should always be written to always drain the channel, regardless of whether or not we're observing over the whole lifetime of the program. Might perhaps make sense redesign `ToChannel` to make the draining part of the API. I'll think about this a bit.